ppa/archive/management/commands/generate_textcorpus.py

```diff
@@ -2,199 +2,238 @@
+            batch_size = total
 
-@contextmanager
-def logging_disabled(highest_level=logging.CRITICAL):
-    """Quick way to suppress solr logs as we iterate. Taken from https://gist.github.com/simon-weber/7853144"""
-    previous_level = logging.root.manager.disable
-    logging.disable(highest_level)
-    try: yield
-    finally: logging.disable(previous_level)
+        # define a generator to iterate solr with
+        iterr = (
+            result
+            for step in range(0, total, batch_size)
+            for result in qset[step : step + batch_size]
+        )
 
+        # if progress bar wanted, tack one on
+        if self.progress:
+            iterr = progressbar(iterr, max_value=total)
 
+        # yield from this generator, progress bar or no
+        yield from iterr
 
+    def iter_works(self):
+        """
+        Simply calls `.iter_solr()` with `item_type` as `'work'`
+        """
+        for d in self.iter_solr(item_type="work"):
+            # source does not need to be a list, will only be one
+            assert len(d["source"]) == 1
+            d["source"] = d["source"][0]
 
-class Command(BaseCommand):
-    """Custom manage command to generate a text corpus from text indexed in Solr"""
+            # yield now
+            yield d
 
-    def add_arguments(self, parser):
-        parser.add_argument(
-            "--path", required=True, help="Directory path to save corpus file(s)."
-        )
+    # save pages
+    def iter_pages(self):
+        """
+        Simply calls `.iter_solr()` with `item_type` as `'page'`
+        """
+        yield from self.iter_solr(item_type="page")
 
-        parser.add_argument(
-            "--doc-limit",
-            type=int,
-            default=-1,
-            help="Limit on the number of documents for corpus generation."
-            "The default of -1 considers ALL documents.",
-        )
+    ### saving to file
+    def save_metadata(self):
+        """
+        Save the work-level metadata as a json file
+        """
+        # get the data from solr
+        data = list(self.iter_works())
+
+        # save if not a dry run
+        if not self.is_dry_run:
+            # save json
+            with open(self.path_meta, "w") as of:
+                json.dump(data, of, indent=2, sort_keys=True)
+
+            # save csv
+            with open(self.path_meta_csv, "w", newline="") as csvfile:
+                # fieldnames we already know
+                fieldnames = [x.split(":")[0] for x in self.WORK_FIELDLIST]
+                writer = csv.DictWriter(csvfile, fieldnames=fieldnames)
+                writer.writeheader()
+                writer.writerows(data)
+
+    def save_pages(self):
+        """
+        Save the page-level metadata as a jsonl file
+        """
+        ### save pages
+        if self.is_dry_run:
+            # iterate in place
+            deque(self.iter_pages(), maxlen=0)
+        else:
+            # save to jsonl or jsonl.gz
```

Comment:
What controls whether it is gzipped or not? is it just based on the filepath?

If it's easy, would be nice to have a flag to generate plain jsonl - seems useful for dev at least.

Reply:
Yes, just based on filename— we could either add a `--pages-filename` argument (right now we only have an output directory flag) or a `--uncompressed` argument?

Reply:
Makes more sense to me to make the uncompressed a flag vs filename, I think we want that to stay the same. Shorter is better, what about `--no-zip` or similar?